diem-move/diem-framework/experimental/sources/NFTGallery.move

```diff
@@ -1,127 +1,159 @@
         let index_opt = index_of_token<TokenType>(gallery, token_id);
-
         if (Option::is_none(&index_opt)) {
             0
         } else {
-            let index = Option::extract(&mut index_opt);
-            NFT::balance(Vector::borrow(gallery, index))
+            let token = Vector::borrow(gallery, Option::extract(&mut index_opt));
+            NFT::get_balance(token)
         }
     }
 
+    /// Returns the overall supply for the given token (across this and potentially other galleries),
+    // aborts if token with the given ID is not found.
+    public fun get_token_supply<TokenType: copy + store + drop>(owner: address, token_id: &GUID::ID): u64 acquires NFTGallery {
+        let gallery = &borrow_global<NFTGallery<TokenType>>(owner).gallery;
+        let index_opt = index_of_token<TokenType>(gallery, token_id);
+        assert!(Option::is_some(&index_opt), EID_NOT_FOUND);
+        let token = Vector::borrow(gallery, Option::extract(&mut index_opt));
+        NFT::get_supply(token)
+    }
+
+    /// Returns a copy of the token content uri
+    public fun get_token_content_uri<TokenType: copy + store + drop>(owner: address, token_id: &GUID::ID): vector<u8> acquires NFTGallery {
+        let gallery = &borrow_global<NFTGallery<TokenType>>(owner).gallery;
+        let index_opt = index_of_token<TokenType>(gallery, token_id);
+        assert!(Option::is_some(&index_opt), EID_NOT_FOUND);
+        let token = Vector::borrow(gallery, Option::extract(&mut index_opt));
+        NFT::get_content_uri(token)
+    }
+
+    /// Returns a copy of the token metadata
+    public fun get_token_metadata<TokenType: copy + store + drop>(owner: address, token_id: &GUID::ID): TokenType acquires NFTGallery {
+        let gallery = &borrow_global<NFTGallery<TokenType>>(owner).gallery;
+        let index_opt = index_of_token<TokenType>(gallery, token_id);
+        assert!(Option::is_some(&index_opt), EID_NOT_FOUND);
+        let token = Vector::borrow(gallery, Option::extract(&mut index_opt));
+        NFT::get_metadata(token)
+    }
+
+    /// Returns a copy of the token parent id
+    public fun get_token_parent_id<TokenType: copy + store + drop>(owner: address, token_id: &GUID::ID): Option<GUID::ID> acquires NFTGallery {
+        let gallery = &borrow_global<NFTGallery<TokenType>>(owner).gallery;
+        let index_opt = index_of_token<TokenType>(gallery, token_id);
+        assert!(Option::is_some(&index_opt), EID_NOT_FOUND);
+        let token = Vector::borrow(gallery, Option::extract(&mut index_opt));
+        NFT::get_parent_id(token)
+    }
+
     /// Transfer `amount` of token with id `GUID::id(creator, creation_num)` from `owner`'s
     /// balance to `to`'s balance. This operation has to be done by either the owner or an
     /// approved operator of the owner.
-    public(script) fun transfer_token_between_galleries<TokenType: store>(
+    public(script) fun transfer_token_between_galleries<TokenType: copy + store + drop>(
```

Comment:
this api looks weird, ideally we should take a GUID::ID directly?

Reply:
Yea, that's because the script functions don't support struct arguments.

Reply:
I just realize that GUID::ID is not a single hash value but two fields